aws-datastore/src/main/java/com/amplifyframework/datastore/network/DataHydration.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.datastore.network;
+
+import com.amplifyframework.core.ResultListener;
+import com.amplifyframework.core.model.Model;
+import com.amplifyframework.datastore.storage.LocalStorageAdapter;
+import com.amplifyframework.datastore.storage.StorageItemChange;
+
+import io.reactivex.Completable;
+import io.reactivex.CompletableEmitter;
+import io.reactivex.schedulers.Schedulers;
+
+/**
+ * "Hydrates" the local DataStore, using model metadata receive from the
+ * {@link RemoteModelState}. Hydration refers to populating the local storage
+ * with values from a remote system.
+ */
+@SuppressWarnings("unused")
+final class DataHydration {
+    private final RemoteModelState remoteModelState;
+    private final LocalStorageAdapter localStorageAdapter;
+
+    DataHydration(
+            RemoteModelState remoteModelState,
+            LocalStorageAdapter localStorageAdapter) {
+        this.remoteModelState = remoteModelState;
+        this.localStorageAdapter = localStorageAdapter;
+    }
+
+    /**
+     * The task of hydrating the DataStore either succeeds (with no return value),
+     * or it fails, with an explanation.
+     * @return An Rx {@link Completable} which can be used to perform the operation.
+     */
+    Completable hydrate() {
+        // Observe the remote model states,
+        return remoteModelState.observe()
+            .subscribeOn(Schedulers.io())
+            .observeOn(Schedulers.io())
+            // For each model state, provide it as an input to a completable operation
+            .flatMapCompletable(modelWithMetadata -> Completable.mergeArray(
+                // And save both the item and the metadata
+                saveItemToStorage(modelWithMetadata), saveMetadataToStorage(modelWithMetadata)
```

Comment:
Maybe I'm overthinking this or don't understand the mechanism but would we have to worry about perhaps the item failing to save but the metadata succeeding and then the item can never be updated because according to the metadata table it thinks it has the latest version?

Reply:
I changed the logic so that the metadata will only be updated if the data write succeeds

Reply:
Update: I changed this so that the saves are chained. I only try to write the metadata if the data saves successfully.